controller_interface/include/controller_interface/async_function_handler.hpp

```diff
@@ -0,0 +1,177 @@
+// Copyright 2024 PAL Robotics S.L.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef CONTROLLER_INTERFACE__ASYNC_FUNCTION_HANDLER_HPP_
+#define CONTROLLER_INTERFACE__ASYNC_FUNCTION_HANDLER_HPP_
+
+#include <atomic>
+#include <condition_variable>
+#include <memory>
+#include <mutex>
+#include <stdexcept>
+#include <string>
+#include <thread>
+#include <vector>
+
+#include "lifecycle_msgs/msg/state.hpp"
+#include "rclcpp/duration.hpp"
+#include "rclcpp/time.hpp"
+#include "rclcpp_lifecycle/state.hpp"
+
+namespace ros2_control
+{
+template <typename T>
+class AsyncFunctionHandler
+{
+public:
+  AsyncFunctionHandler() = default;
+
+  ~AsyncFunctionHandler() { preempt_async_update(); }
+
+  void init(
+    std::function<const rclcpp_lifecycle::State &()> get_state_function,
+    std::function<T(const rclcpp::Time &, const rclcpp::Duration &)> async_function)
+  {
+    if (get_state_function == nullptr || async_function == nullptr)
+    {
+      throw std::runtime_error("AsyncFunctionHandler parsed functions are not valid!");
+    }
+    if (thread_.joinable())
+    {
+      throw std::runtime_error("AsyncFunctionHandler already initialized!");
+    }
+    get_state_function_ = get_state_function;
+    async_function_ = async_function;
+  }
+
+  T trigger_async_update(const rclcpp::Time & time, const rclcpp::Duration & period)
+  {
+    initialize_async_update_thread(get_state_function_, async_function_);
+    std::unique_lock<std::mutex> lock(async_mtx_, std::try_to_lock);
+    if (!async_update_ready_ && lock.owns_lock())
+    {
+      async_update_ready_ = true;
```

Comment:
I'd switch around the meaning of `async_update_ready_`. For me, the update being ready means is done and can be called again.

Reply:
Done, I've renamed the variable